services/com.mde-network.ep.toolfunctions.xtext/src/main/java/com/mdenetnetwork/ep/toolfunctions/xtext/XtextTool.java

```diff
@@ -0,0 +1,227 @@
+import java.nio.file.Path;
+import java.nio.file.StandardOpenOption;
+import java.util.List;
+import java.util.zip.ZipEntry;
+import java.util.zip.ZipOutputStream;
+import java.io.FileOutputStream;
+import java.io.IOException;
+
+import com.google.gson.JsonObject;
+import com.google.gson.JsonParser;
+import com.google.gson.JsonPrimitive;
+
+import org.eclipse.emf.ecore.resource.Resource;
+import org.eclipse.emf.ecore.xmi.impl.EcoreResourceFactoryImpl;
+import org.eclipse.xtext.xtext.wizard.cli.CliProjectsCreatorMain;
+import org.apache.hc.client5.http.HttpHostConnectException;
+import org.apache.hc.client5.http.entity.mime.MultipartEntityBuilder;
+import org.apache.hc.core5.http.ContentType;
+import org.apache.hc.core5.http.HttpEntity;
+import org.apache.hc.client5.http.entity.mime.HttpMultipartMode;
+import org.apache.hc.client5.http.fluent.Request;
+
+public class XtextTool  { 
+
+	static final String PROJECT_PATH = "./xtext-project/" ;
+	static final String PROJECT_FILENAME = "xtext-project.zip" ;
+	static final String PROJECT_FILE_PATH = "./" + PROJECT_FILENAME ;
+	static final String ES_JSON_RESPONSE_FIELD = "editorUrl";
+	static final String ES_JSON_RESPONSE_FIELD_STATUS = "editorStatusUrl";
+	static final int FILE_BUFFER_SIZE = 1024;
+	
+	public XtextTool() {
+		
+		// Register resource factories
+		Resource.Factory.Registry.INSTANCE.getExtensionToFactoryMap().put("ecore", new EcoreResourceFactoryImpl());
+	}
+
+
+	public void run(String languageName, String baseName, String extension, String grammar, List<ProjectFile> projectFiles, OutputStream outputStream, JsonObject response) throws Exception {
+
+
+		
+		// Create Xtext project with web editor enabled
+		File projectFolder = new File(PROJECT_PATH);
+		
+		String[] args = new String[] {
+				"-targetDir=" + projectFolder.getAbsolutePath(),
+				"-languageName="+languageName,
+				"-baseName="+baseName,
+				"-extension="+extension,
+				"-enabledProjects=UIProject,GenericIDE,RuntimeTest,UITest,Web",
+				"-buildSystem=MAVEN",
+				"-sourceLayout=PLAIN",
+				"-languageServer=FATJAR",
+				"-xtextVersion=2.30.0",
+				"-projectLayout=FLAT",
+				"-javaVersion=JAVA11"
+			};
+
+		
+		CliProjectsCreatorMain.main(args);
+		
+		
+		
+		
+		// Add activity files
+		if (grammar != null) {
+			final String grammarSrcPath = languageName.replace('.', '/');
+			final String xtextGrammarPath= PROJECT_PATH + baseName + "/src/" + grammarSrcPath  + ".xtext";
+			recreateGrammarFile(xtextGrammarPath, grammar);
+		}
+		
+		//TODO add activity files
+		
+		
+		// Compress files for transfer to build server
+		
+		ZipOutputStream projectFileStream = new ZipOutputStream( new FileOutputStream(PROJECT_FILE_PATH) );
+		
+		compressDirectory(projectFolder, "", projectFileStream);
+		 
+		projectFileStream.flush();
+		projectFileStream.close();
+		 
+		
+		// Send to the build server
+		
+		final String EDITOR_SERVER_URL= "http://127.0.0.1:10001/xtext/upload";  // TODO set URL via environment variable 
```

Comment:
Why is this on a separate server at all?

Reply:
Because of the challenge building maven projects within the google cloud function environment. The separate structure is kept so if in the future the build steps can be made to run inside the restricted function environment. However, it may be more suitable to use something like google cloud build for this if we want to retain as much of the cloud based architecture. 

Even though they are separate they are configured to be on the server/container but we should restructure appropriately as part of the parallel build support of mdenet/educationplatform#81

Reply:
Via unix sockets as urls (I'm not sure that's supported by java http library being used)? or removing the http request entirely? 

If we choose the second option it would probably be best to make this functionality a command line application and add the request handling to the editor express that then invokes it, or, refactor it to be completely to be servelet based including the express part so that it can be deployed on the tomcat server.  

However, if we decide to use google cloud build the project generation part would be moved to a build container and have a cloud function that invokes the build.

Reply:
Added to the tracking issue and added details to https://github.com/mdenet/educationplatform/issues/84, to be resolved at a later date.